### ISSUE TITLE
"Could not find a valid mapping" Error on Update

### ISSUE THREAD

**Issue by jamilabreu** [neo4jrb/devise-neo4j#16]
When updating a person in my Person model on a clean rails app, I'm getting the following error:

```
Could not find a valid mapping for #<Person created_at:...
```

The changes get saved, but the error comes up. Here's part of the trace:

```
devise (3.4.1) lib/devise/mapping.rb:43:in `find_scope!'
devise (3.4.1) lib/devise/controllers/helpers.rb:166:in `signed_in_root_path'
devise (3.4.1) app/controllers/devise/registrations_controller.rb:132:in `after_update_path_for'
devise (3.4.1) app/controllers/devise/registrations_controller.rb:62:in `update'
actionpack (4.2.1) lib/action_controller/metal/implicit_render.rb:4:in `send_action'
actionpack (4.2.1) lib/abstract_controller/base.rb:198:in `process_action'
actionpack (4.2.1) lib/action_controller/metal/rendering.rb:10:in `process_action'
actionpack (4.2.1) lib/abstract_controller/callbacks.rb:20:in `block in process_action'
activesupport (4.2.1) lib/active_support/callbacks.rb:117:in `call'
activesupport (4.2.1) lib/active_support/callbacks.rb:117:in `call'
activesupport (4.2.1) lib/active_support/callbacks.rb:555:in `block (2 levels) in compile'
```
Update: only getting this error when attempting to update via Devise::RegistrationsController#update

**Comment by andrewhavens** [comment]
I am also running into this issue. I've implemented a custom token auth which calls the devise `sign_in` helper:

```ruby
def authenticate_user_from_token!
    authenticate_with_http_token do |token, options|
      user = User.find_by(email: options[:email])
      if user && Devise.secure_compare(user.authentication_token, token)
        sign_in user, store: false
      end
    end
  end
```

It's failing with the same message because [the `sign_in` helper is also calling `find_scope!`](https://github.com/plataformatec/devise/blob/master/lib/devise/controllers/sign_in_out.rb#L32). In my case, I've traced it down to [this line](https://github.com/plataformatec/devise/blob/master/lib/devise/mapping.rb#L41):

```ruby
Devise.mappings.each_value { |m| return m.name if obj.is_a?(m.to) }
```

The comparison of `obj.is_a?(m.to)` is failing because `obj` is a `User`, but `m.to` is returning: 

```
User(authentication_token: Object, confirmation_sent_at: DateTime, confirmation_token: Object, confirmed_at: DateTime, created_at: DateTime, current_sign_in_at: DateTime, current_sign_in_ip: Object, email: Object, encrypted_password: Object, facebook_token: Object, first_name: Object, last_name: Object, last_sign_in_at: DateTime, last_sign_in_ip: Object, remember_created_at: DateTime, reset_password_sent_at: DateTime, reset_password_token: Object, sign_in_count: Object, tos: Object, unconfirmed_email: Object, updated_at: DateTime)
```

I traced this down to find why `m.to` was returning this. `m` is an instance of a `Devise::Mapping`. The `#to` method simply calls `@klass.get`. When the mapping is initialized, `@klass` is set to:

```ruby
@class_name = User # in my case
@klass = Devise.ref(@class_name)
```

[When `Devise.ref` is executed, it calls this code](https://github.com/plataformatec/devise/blob/master/lib/devise.rb#L285-L303), which I think is the core issue:

```ruby
  def self.ref(arg)
    if defined?(ActiveSupport::Dependencies::ClassCache)
      ActiveSupport::Dependencies::reference(arg)
      Getter.new(arg)
    else
      ActiveSupport::Dependencies.ref(arg)
    end
  end
```

...which then calls `Devise::Getter.new(User)` which results in:

```
#<Devise::Getter:0x007f98d95c0ef8 @name=User(authentication_token: Object, confirmation_sent_at: DateTime, confirmation_token: Object, confirmed_at: DateTime, created_at: DateTime, current_sign_in_at: DateTime, current_sign_in_ip: Object, email: Object, encrypted_password: Object, facebook_token: Object, first_name: Object, last_name: Object, last_sign_in_at: DateTime, last_sign_in_ip: Object, remember_created_at: DateTime, reset_password_sent_at: DateTime, reset_password_token: Object, sign_in_count: Object, tos: Object, unconfirmed_email: Object, updated_at: DateTime)>
```

I don't know the solution to this yet. I'm wondering if it has anything to do with class caching. I just upgraded to v5 of the neo4j gem. I don't think I remember seeing this before.

**Comment by subvertallchris** [comment]
We solved https://github.com/neo4jrb/neo4j/issues/717 the other day, which had very similar symptoms: an object was given to a method that checked its type but it claimed the type did not match, even when it should have.

In that case, it was the result of class caching. I worked around it by getting rid of the `is_a?` call and using other qualities of the models, but I didn't spend a lot of time trying to get to the root cause. I wonder if https://github.com/neo4jrb/neo4j/blob/master/lib/neo4j/active_node/labels.rb#L8 is contributing to it. It might be hanging onto old versions of classes after the Rails reloader has instantiated new ones.

**Comment by subvertallchris** [comment]
If one of you guys have a chance, can you see if https://github.com/neo4jrb/neo4j/pull/855 fixes it? Just `gem 'neo4j', github: 'neo4jrb/neo4j', branch: 'constantize_problem'`

**Comment by andrewhavens** [comment]
@subvertallchris This might have worked. I waited until I started seeing the errors again, updated my Gemfile to your branch, bundled, then restarted the app and the errors went away. However, restarting the app might have fixed this as well. I'll keep running your branch over the next few days and see if the error comes back.

**Comment by subvertallchris** [comment]
If it's the same root cause, the restart will always clear up the
temporarily. Hopefully, this does it permantnetly, but even if it doesn't,
I think we'll have an easier time figuring out what's going on. Thanks for
keeping me posted! I'll keep my fingers crossed.

**Comment by subvertallchris** [comment]
I merged the PR into Neo4j and will delete the branch soon, so you can use neo4j's master or lock at ref 05abc109c73740e3ed10216aecd54405b5047383.

**Comment by djvs** [comment]
Same issue, crops up after some time, restart clears up.

**Comment by subvertallchris** [comment]
Does updating to the current branch of neo4j resolve it?

**Comment by djvs** [comment]
You mean the master branch?

**Comment by djvs** [comment]
Seems so yes

**Comment by subvertallchris** [comment]
Seems to fix it? And yeah, master branch will do it but it's really the
commit in the PR linked in this issue. It was merged in the other night.

**Comment by thermistor** [comment]
Did not fix it for me. I am running neo4j 5.0.5 and devise 3.5.1 and I still got this error.

I am using a `to_prepare` clause in `neo4j/railtie.rb` and that appears to have solved the issue for me:

https://github.com/thermistor/neo4j/commit/a2a31f1826c47a4a28d6c918bf3c7adbe09c394a

Happy to make a pull request if someone could weigh in on whether this is the right approach.

**Comment by subvertallchris** [comment]
This will be run before every request in development, I'm not sure that's
ideal, but I'm also not a Rails expert so maybe this is how things are
done. It sounds like there's another reload event occurring that isn't
triggering the callback I added to the gem?

**Comment by thermistor** [comment]
I'll see if I can create a little project that duplicates the issue later this week. I was able to cause the problem pretty easily by simple changing code in my devise model, User.rb, and then doing any action that would call Devise::Mapping.find_scope!

**Comment by subvertallchris** [comment]
@thermistor 5.0.5 doesn't actually include the fix for this, you'll need to stick with master or the PR I referenced for now.

**Comment by thermistor** [comment]
@subvertallchris Ah ok thanks, bad assumption on my part.

**Comment by subvertallchris** [comment]
This fix was worked into neo4j 5.0.15, so I'm going to close the issue.
